packages/transport-native-ble/src/api/nativeBleManager.ts

```diff
@@ -0,0 +1,438 @@
+/* eslint-disable require-await */
+import {
+    BleError,
+    BleErrorCode,
+    BleManager,
+    Characteristic,
+    Device,
+    LogLevel,
+    ScanOptions,
+} from 'react-native-ble-plx';
+
+const DEBUG_LOGS = true;
+
+const debugLog = (...args: any[]) => {
+    if (DEBUG_LOGS) {
+        // eslint-disable-next-line no-console
+        console.log('NativeBleManager: ', ...args);
+    }
+};
+
+const bluetoothInfoCache: { [deviceUuid: string]: any } = {}; // Allows us to give more granulary error messages.
+
+let connectOptions: Record<string, unknown> = {
+    requestMTU: 247,
+    connectionPriority: 1,
+};
+
+/**
+ * Returns the instance of the Bluetooth Low Energy Manager. It initializes it only
+ * when it's first needed, preventing the permission prompt happening prematurely.
+ * Important: Do NOT access the _bleManager variable directly.
+ * Use this function instead.
+ * @returns {BleManager} - The instance of the BleManager.
+ */
+let _bleManager: BleManager | null = null;
+export const bleManagerInstance = (): BleManager => {
+    if (!_bleManager) {
+        _bleManager = new BleManager();
+        _bleManager.setLogLevel(LogLevel.Verbose);
+    }
+
+    return _bleManager;
+};
+
+export const NUS_SERVICE_UUID = '6e400001-b5a3-f393-e0a9-e50e24dcca9e';
+// const NUS_CHARACTERISTIC_NOTIFY = '6e400003-b5a3-f393-e0a9-e50e24dcca9e';
+// const NUS_CHARACTERISTIC_TX = '6e400002-b5a3-f393-e0a9-e50e24dcca9e';
+export const devicesUUIDs: Array<string> = [NUS_SERVICE_UUID];
+
+const READ_TIMEOUT_MS = 3 * 60 * 1000; // 3 minutes
+const READ_FREQUENCY_MS = 10;
+
+const scanOptions: ScanOptions = {};
+
+type Base64String = string;
+type DeviceMessage = { value: Base64String; timestamp: number };
+
+export interface BLEDeviceWithMetadata {
+    bleDevice: Device;
+    writeUuid: string;
+    notifyUuid: string;
+    readOutputQueue: DeviceMessage[];
+}
+
+export interface BLEScannedDevice {
+    lastSeenTimestamp: number;
+    bleDevice: Device;
+}
+
+class NativeBleManager {
+    bleManager: BleManager;
+    private devicesScanList: BLEScannedDevice[] = [];
+    public onconnect?: (device: Device) => void;
+    public ondisconnect?: (device: Device) => void;
+
+    // There could be devices that are connected to system but not to our app
+    private appConnectedDevices: BLEDeviceWithMetadata[] = [];
+
+    constructor() {
+        this.bleManager = bleManagerInstance();
+        this.enumerate();
+    }
+
+    private async enumerate() {
+        const connectedDevices = await bleManagerInstance().connectedDevices(devicesUUIDs);
```

Comment:
Can not be `this.bleManager` accessed instead? Since the `_bleManager` should be initialized only once, it should be safe.

```suggestion
        const connectedDevices = this.bleManager.connectedDevices(devicesUUIDs);
```